crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1340,21 +1340,62 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the length value of a type.
+    ///
+    /// This is used to determine the value that would be returned
+    /// when `len(x)` is called on an object `x`.
+    fn len(&self, db: &'db dyn Db) -> Len {
+        match self {
+            Type::StringLiteral(string) => string.as_str(db).chars().count().into(),
+            Type::BytesLiteral(bytes) => bytes.len(db).into(),
+
+            Type::Tuple(tuple) => {
+                let elements = tuple.elements(db);
+
+                if elements.iter().any(|it| matches!(it, Type::Todo(..))) {
+                    Len::StaticallyUnknown
+                } else {
+                    elements.len().into()
+                }
+            }
+
+            Type::Never => Len::Never,
+            Type::BooleanLiteral(_) => Len::Never,
+            Type::IntLiteral(_) => Len::Never,
+
+            // Likely unreachable
+            Type::SliceLiteral(_) => Len::Never,
+
+            Type::Instance(instance) => match instance.class.known(db) {
+                Some(KnownClass::Bool) => Len::Never,
+                _ => Len::StaticallyUnknown,
+            },
+
+            _ => Len::StaticallyUnknown,
+        }
+    }
+
     /// Return the outcome of calling an object of this type.
     #[must_use]
     fn call(self, db: &'db dyn Db, arg_types: &[Type<'db>]) -> CallOutcome<'db> {
         match self {
             // TODO validate typed call arguments vs callable signature
-            Type::FunctionLiteral(function_type) => {
-                if function_type.is_known(db, KnownFunction::RevealType) {
-                    CallOutcome::revealed(
-                        function_type.signature(db).return_ty,
-                        *arg_types.first().unwrap_or(&Type::Unknown),
-                    )
-                } else {
-                    CallOutcome::callable(function_type.signature(db).return_ty)
+            Type::FunctionLiteral(function_type) => match function_type.known(db) {
+                Some(KnownFunction::RevealType) => CallOutcome::revealed(
+                    function_type.signature(db).return_ty,
+                    *arg_types.first().unwrap_or(&Type::Unknown),
+                ),
+
+                Some(KnownFunction::Len) => {
+                    let [only_arg] = arg_types else {
+                        return CallOutcome::callable(Len::StaticallyUnknown.as_type(db));
+                    };
+
+                    CallOutcome::callable(only_arg.len(db).as_type(db))
```

Comment:
We fall back to a hard-coded `int` here in case we can't statically compute a more precise type. I think I would prefer a slightly different logic here where we try to compute a statically-know result type, but if that fails, fall back to normal signature checking, like we do in the catchall `_ =>` branch below. This would lead to [the same result](https://github.com/python/typeshed/blob/02533d13ea43423fab76f6e8d2decf1350fcc302/stdlib/builtins.pyi#L1480) under normal circumstances, but would also handle cases where we have a custom `typeshed`, for example.